symmetric-core/src/main/java/org/jumpmind/symmetric/service/impl/DataService.java

```diff
@@ -3596,94 +3597,152 @@ protected void convertDataToReload(Data data, Table table, TriggerHistory hist,
+        long queryStartDataId = minDataId;
+        long queryEndDataId = Long.min(minDataId + RECAPTURE_DATA_COMMIT_LIMIT - 1, maxDataId);
+        List<Data> dataList = findData(queryStartDataId, queryEndDataId);
+        if (dataList == null || dataList.size() < 1) {
+            return 0;
+        }
+        int reCapturedCount = 0;
+        while (queryStartDataId <= maxDataId) {
+            if (dataList.size() > 0) {
+                reCapturedCount += reCaptureData(dataList);
+            }
+            queryStartDataId = ++queryEndDataId;
+            queryEndDataId = Long.min(queryEndDataId + RECAPTURE_DATA_COMMIT_LIMIT - 1, maxDataId);
+            if (queryStartDataId > maxDataId) {
+                break;
+            }
+            dataList = findData(queryStartDataId, queryEndDataId);
         }
-        return count;
+        return reCapturedCount;
     }
 
+    /**
+     * Attempts to recapture stale data rows and re-insert them back into sym_data with a new transaction ID (all records are in one transaction).
+     * 
+     * @return number of recaptured data rows
+     */
     protected int reCaptureData(List<Data> dataList) {
         List<Data> insertList = new ArrayList<Data>();
-        ISqlTransaction transaction = null;
-        Table table = null;
-        String[] keys = null;
-        Data lastData = null;
         long lastUpdateTimestamp = 0;
         String recaptureTransactionId = this.generateTransactionIdFromTimestamp("recapture-", Instant.now());
+        int insertedCount = 0;
         try {
             for (Data data : dataList) {
-                lastData = data;
-                if (data.isPreRouted() || !(data.getDataEventType().isDml())) {
+                Data recapturedData = fetchRecapturedData(data, recaptureTransactionId);
+                if (recapturedData == null) {
                     continue;
                 }
-                TriggerHistory hist = data.getTriggerHistory();
-                Set<TriggerRouter> triggerRouters = engine.getTriggerRouterService().getTriggerRouterForTableForCurrentNode(
-                        hist.getSourceCatalogName(), hist.getSourceSchemaName(), hist.getSourceTableName(), false);
-                table = platform.getTableFromCache(hist.getSourceCatalogName(), hist.getSourceSchemaName(), hist.getSourceTableName(), false);
-                if (triggerRouters != null && triggerRouters.size() > 0 && table != null) {
-                    Trigger trigger = triggerRouters.iterator().next().getTrigger();
-                    table = table.copyAndFilterColumns(hist.getParsedColumnNames(), hist.getParsedPkColumnNames(), true, false);
-                    if (data.getDataEventType() == DataEventType.INSERT) {
-                        keys = data.toParsedRowData();
-                        if (keys != null && keys.length >= table.getPrimaryKeyColumnCount()) {
-                            keys = ArrayUtils.subarray(keys, 0, table.getPrimaryKeyColumnCount());
```

Comment:
Are primary keys ALWAYS in the front in sym_data.row_data?

Reply:
Yes, it orders the columns so that primary keys are at the front of row_data.